### DISCUSSION TITLE
How to specify a default Jupyter kernel via configuration?

### DISCUSSION THREAD

**Discussion by jtkiley** [microsoft/vscode-jupyter#13032] [Questions and Answers] [answered]
I am trying to set a kernel via some kind of version-controllable configuration (in a devcontainer, if that matters), so I don't have to respecify kernels every time I rebuild (or when a new user spins up the a container).

Here is my most promising lead so far (hence why I'm asking here):

In the docs (see [here](https://github.com/microsoft/vscode-jupyter/wiki/Jupyter-Kernels-and-the-Jupyter-Extension#kernel-selection)), it says:

>When you pick a kernel in the kernel picker, that kernel will then be associated with that file and automatically selected the next time that you open it. If nothing has been selected, the Jupyter extension will take a look at the notebook metadata and attempt to select the best kernel for the file.

Is there a way to control that association so that a specified kernel will work automatically? I'd like to set it up in my devcontainer, and have it just work.

**Comment by jablka** [comment]
I have the same problem.
VSCode recently started asking for Kernel: "Select Kernel".
It wasn't doing previously... Not sure what happened.
Any suggestion appreciated.

**Comment by jablka** [comment]
update:
The problem occurs if one has more Python installations in the computer.
I've unistalled Anaconda distribution, and the prompt "Select Kernel" disappeared. Now it automatically finds the only one Python.
But this is not a good solution though...

**Comment by pamelafox** [comment]
I'm also trying to figure this out - I was hoping that at least if I override the language version in the metadata in the notebooks, it would auto-select that version, but it still prompts for an interpreter the first time. I filed this issue asking for a project-wide configuration: https://github.com/microsoft/vscode-jupyter/issues/13641

**Comment by jablka** [comment]
I see a similar discussion here: **Add a default kernel option to settings.json** https://github.com/microsoft/vscode/issues/130946

**Comment by Riezebos** [comment]
As a workaround, putting all Python environments except a single one in the setting `"jupyter.kernels.excludePythonEnvironments"` will cause VSCode to auto-select that one as kernel by default. 

Intellisense nicely suggests all the interpreter paths in settings.json, but you might end up with something like:

```json
{
"jupyter.kernels.excludePythonEnvironments": [
    "~/.pyenv/versions/3.10.10/bin/python",
    "/opt/homebrew/bin/python3.10",
    "~/.pyenv/versions/mambaforge/bin/python",
    "~/.pyenv/versions/3.10.2/bin/python",
    "~/.pyenv/versions/3.11.2/bin/python",
    "/opt/homebrew/bin/python3",
    "~/.pyenv/versions/3.11.7/bin/python",
    "/opt/homebrew/bin/python3.12",
    "~/.pyenv/versions/3.7.16/bin/python",
    "~/.pyenv/versions/3.8.16/bin/python",
    "~/.pyenv/versions/3.9.16/bin/python",
    "/opt/homebrew/bin/python3.9",
    "/usr/bin/python3"
  ],
}
```

**Comment by DonJayamanne** [comment, marked as answer]
Please see solution here https://github.com/microsoft/vscode/issues/130946#issuecomment-1899389049
